crux_core/src/capability/executor.rs

```diff
@@ -129,33 +111,47 @@ impl QueuingExecutor {
         }
     }
 
-    fn run_task(&self, task_id: Uuid) {
+    fn run_task(&self, task_id: TaskId) {
         let mut tasks = self.tasks.lock().unwrap();
-        let mut task = tasks.remove(&task_id).unwrap();
+        let mut task = tasks
+            .get_mut(*task_id as usize)
+            .and_then(|task| task.take())
+            .expect("Task is missing");
```

Comment:
Isn't this susceptible to the concurrency problem? This could mean that the task is being processed by anoter thread, right? Or do I still have a poor conceptual model of this? 😅

Reply:
You are right but I don't see how to avoid it short of putting a mutex around the task. Which we could do - but there are so many nested mutexes, I think we should have a look at the multithreading strategy in general first. I'm worried that with so many locks everywhere, if we ever tried to multithread things we'd get a deadlock anyway. And AFAICT mutlithreading is not tested anywhere.

Reply:
I think the sensible thing to do with that is just skipping the task - it's already being executed. I'm not sure if it's possible for the task to be woken _while_ it's executing (when it's not in the slab), is it? This needs a bit of thinking but it seems safe at first blush to just ignore the task if the future is `None`.

In terms of multithreading - if the locks are kept reasonably granular, it should be fine. It is not really tested, partly because it only really manifests when inside a shell. The key property of this though, is that none of the work is particularly heavy, so even if we block the other threads for the entire duration of `run_all` we should still be totally fine. Until someone puts an LLM training job in a task that is... 😅

Reply:
I don't think you can ignore the wake notification if the task is being executed - because it won't get woken again and won't complete. 

On the other hand, is it actually possible to receive a wake notification whilst being executed... not certain but I think in practice 'No'. The polling mechanism will continue until a task is 'pending' (e.g. needs effect to run on the shell), then registers a callback with the shell and suspend with `Poll::Pending`. Only once the shell resolves will the notification be sent, by which time the future will be back in its slot. I suppose in theory there is a race condition here but it's impossibly unlikely to be hit.

One could write a malicious future which subverted this.. but we won't be doing that.

Reply:
I'm reasonably certain we can't leave the panic in because one day, that will crash an app somewhere without warning. For one, shell requests are not the only reason tasks wake, channels also wake them (although this may be through a different mechanism, can't quite say with certainty 🤔 ). The other case, which I'm a lot more certain about is subscription requests, which can be resolved repeatedly, and with no throttling, so it could easily be the case that while the task is executing on one thread, it is also resolved on another by a subsequent event on the same subcription, and 💥 . 

However - I think we can ignore the missing tasks _and_ avoid the problem of task not getting woken up if we miss the chance, by keeping a separate register of tasks which should be woken up when possible. That way we can easily ignore the task if it's missing (because it's executing), but we still retain the fact that it's ready to go again when it can. Once the original execution finishes, it will review the ready register and execute the task again.

Reply:
One way that would not require much more plumbing would be to requeue the notification and continue looping. Yes this is a busy-loop but as you say we don't expect any one event to take very long

Reply:
Hmmm, I think that's actually equivalent to the registry anyway. If the execution plan is 
* pull off ready queue, mark as ready
* run everything marked as ready - if the task isn't available skip
* repeat

That will be a busy loop as well, right? However, if we keep the idea of `did_do_work`, then I think that actually solves the whole thing, because the skipped tasks wouldn't mark that as true, so the loop would stop, but the other thread's loop would then pick them up, if I'm picturing this correctly

Reply:
Either way, we should add a multi-threaded test for this sort of thing, but maybe in another PR. Something along the lines of
* App which starts a subscription and collects all the arriving things in its model
* Run the app, and start several threads which all spam the subscription with numbers
* Verify that all the numbers have arrived

For bonus points we could put some kind of busy loop into the subscription futures to make the tasks run longer each time